circup/__init__.py

```diff
@@ -409,6 +410,616 @@ def __repr__(self):
+        """
+        url = urlparse(device_path)
+        auth = HTTPBasicAuth("", url.password)
+        r = requests.delete(module_path, auth=auth)
+        r.raise_for_status()
+
+    def update(self, module):
+        """
+        Delete the module on the device, then copy the module from the bundle
+        back onto the device.
+
+        The caller is expected to handle any exceptions raised.
+        """
+        self._update_http(module)
+
+    def _update_http(self, module):
+        """
+        Update the module using web workflow.
+        """
+        if module.file:
+            # Copy the file (will overwrite).
+            self.install_file_http(module.bundle_path, module.path)
+        else:
+            # Delete the directory (recursive) first.
+            url = urlparse(module.path)
+            auth = HTTPBasicAuth("", url.password)
+            r = requests.delete(module.path, auth=auth)
+            r.raise_for_status()
+            self.install_dir_http(module.bundle_path, module.path)
+
+
+def _get_modules_file(path):
+    """
+    Get a dictionary containing metadata about all the Python modules found in
+    the referenced file system path.
+
+    :param str path: The directory in which to find modules.
+    :return: A dictionary containing metadata about the found modules.
+    """
+    result = {}
+    if not path:
+        return result
+    single_file_py_mods = glob.glob(os.path.join(path, "*.py"))
+    single_file_mpy_mods = glob.glob(os.path.join(path, "*.mpy"))
+    package_dir_mods = [
+        d
+        for d in glob.glob(os.path.join(path, "*", ""))
+        if not os.path.basename(os.path.normpath(d)).startswith(".")
+    ]
+    single_file_mods = single_file_py_mods + single_file_mpy_mods
+    for sfm in [f for f in single_file_mods if not os.path.basename(f).startswith(".")]:
+        metadata = extract_metadata(sfm)
+        metadata["path"] = sfm
+        result[os.path.basename(sfm).replace(".py", "").replace(".mpy", "")] = metadata
+    for package_path in package_dir_mods:
+        name = os.path.basename(os.path.dirname(package_path))
+        py_files = glob.glob(os.path.join(package_path, "**/*.py"), recursive=True)
+        mpy_files = glob.glob(os.path.join(package_path, "**/*.mpy"), recursive=True)
+        all_files = py_files + mpy_files
+        # default value
+        result[name] = {"path": package_path, "mpy": bool(mpy_files)}
+        # explore all the submodules to detect bad ones
+        for source in [f for f in all_files if not os.path.basename(f).startswith(".")]:
+            metadata = extract_metadata(source)
+            if "__version__" in metadata:
+                metadata["path"] = package_path
+                result[name] = metadata
+                # break now if any of the submodules has a bad format
+                if metadata["__version__"] == BAD_FILE_FORMAT:
+                    break
+    return result
+
+
+class USBBackend(Backend):
```

Comment:
I'd call this `DiskBackend` and have it take in the folder location. That leaves a world where we add a USB (or not) serial backend for circup over serial. It also covers the case where it is operating on a separate folder on the same disk.